### PR TITLE
feat(multitenancy): add support for multitenancy and handle the same in router, producer, consumer, drainer and analytics

### PR DESCRIPTION
**Pull request by jagan-jaya**
## Type of Change

- [ ] Bugfix
- [X] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

1. Add tenant id in sessionState and populate it from request header. All component should take tenant_id only from sessionState
2. Prefix_tenant_id with redis key. Separate connection management for redis is not needed
3. Add redirection url based on the tenant base_url
4. Process tracker to handle tenant-id. To be handled in producer, consumer and drainer

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

 To establish tenancy in our application and start using the default tenant as 'Hyperswitch' and also offer a 'test' tenant in all environments

## How did you test it?

**Test for Scheduler**
1. Enable multitenancy by changing [multitenancy] in development.toml
2. Create multiple tenants by adding [multitenancy.tenants] in development.toml
3. Start Router, Producer and Consumer
4. Create a payment by passing `x-tenant-id` as `<your-tenant>` in the request header in localhost:8080
5. Do the same for both tenants and verify if the payment related data are stored in the respective schema of each tenant
6. Simultaneously each payment should have an process tracker entry in individual schema
7. Verify if the process tracker status for both task is changing to terminal state in DB indicating that producer and consumer are able to pick up the task from different schema and update the status respectively

**Test for Drainer**
1. Enable multitenancy by changing [multitenancy] in development.toml
2. Create multiple tenants by adding [multitenancy.tenants] in development.toml
3. Start Router and Drainer `cargo r --bin drainer`
4. Create a merchant by passing `x-tenant-id` as `<your-tenant>` in the request header in localhost:8080
5. Enable KV for the merchant using below curl
```curl --location 'https://localhost:8080/accounts/{{merchant_id}}/kv' \
--header 'Content-Type: application/json' \
--header 'api-key: {{admin_api_key}}' \
--data '{
  "kv_enabled": true
}'
```
6. Create a payment by passing `x-tenant-id` as `<your-tenant>` in the request header in localhost:8080
7. Once the payment get succeeded, make sure payment_intent and payment_attempt table are updated

**Test for Redirection**
1. Enable multitenancy by changing [multitenancy] in development.toml
2. Create multiple tenants by adding [multitenancy.tenants] in development.toml
3. Start Router and make a 3DS payment, the redirection url should be correct as per the tenant configuration

## Checklist

- [ ] I formatted the code `cargo +nightly fmt --all`
- [ ] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/services/api.rs [hunk 1]
```diff
@@ -1001,17 +1004,54 @@ where
...
+                    message: "Missing tenant Id".to_string(),
+                }
+                .switch()
+            })
+            .map(|tenant_id| {
+                if !state
+                    .conf
+                    .multitenancy
+                    .get_tenant_names()
+                    .contains(&tenant_id.to_string())
```
Comment (NishantJoshi00) [inline crates/router/src/services/api.rs:1031]: This is more expensive compared to a hash-map lookup

**Review by NishantJoshi00 — CHANGES_REQUESTED** [review]

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/db/api_keys.rs [hunk 2]
```diff
@@ -527,15 +527,12 @@ mod tests {
...
-                    merchant_id,
-                    hashed_api_key.into_inner()
-                ),)
-                .await
-                .is_none()
-        )
+        assert!(ACCOUNTS_CACHE
+            .get_val::<storage::ApiKey>(CacheKey {
+                key: format!("{}_{}", merchant_id, hashed_api_key.into_inner()),
+                prefix: String::default(),
```
[the remark this answers is not in the document]
Comment (jagan-jaya) [reply crates/router/src/db/api_keys.rs:533]: Used only in tests, anyway added a Todo task to make it usage specific enum

File: crates/storage_impl/src/redis/cache.rs [hunk 3]
```diff
@@ -126,6 +126,22 @@ pub struct Cache {
     inner: MokaCache<String, Arc<dyn Cacheable>>,
 }
 
+#[derive(Debug, Clone)]
+pub struct CacheKey {
+    pub key: String,
+    pub prefix: String,
```
[the remark this answers is not in the document]
Comment (jagan-jaya) [reply crates/storage_impl/src/redis/cache.rs:132]: Added todo to create usage specific enum.

File: crates/scheduler/src/consumer.rs [hunk 4]
```diff
@@ -125,21 +144,20 @@ pub async fn consumer_operations<T: SchedulerAppState + 'static>(
     let group_name = settings.consumer.consumer_group.clone();
     let consumer_name = format!("consumer_{}", Uuid::new_v4());
 
-    let group_created = &mut state
+    let _group_created = &mut state
         .get_db()
         .consumer_group_create(&stream_name, &group_name, &RedisEntryId::AfterLastID)
         .await;
```
[the remark this answers is not in the document]
Comment (jagan-jaya) [reply crates/scheduler/src/consumer.rs:150]: Those are redundant and insignificant logs.

File: crates/router/src/types/storage/payment_attempt.rs [hunk 5]
```diff
@@ -124,8 +124,11 @@ mod tests {
             ..PaymentAttemptNew::default()
         };
 
-        let response = state
-            .store
+        let store = state
+            .stores
+            .get(state.conf.multitenancy.get_tenant_names().first().unwrap())
```
[the remark this answers is not in the document]
Comment (jagan-jaya) [reply crates/router/src/types/storage/payment_attempt.rs:129]: used only in tests

File: crates/router/src/types/storage/payment_attempt.rs [hunk 6]
```diff
@@ -124,8 +124,11 @@ mod tests {
...
+            .unwrap();
```
[the remark this answers is not in the document]
Comment (jagan-jaya) [reply crates/router/src/types/storage/payment_attempt.rs:130]: used only in tests

File: crates/router/src/types/storage/payment_attempt.rs [hunk 7]
```diff
@@ -165,14 +168,16 @@ mod tests {
             attempt_id: attempt_id.clone(),
             ..PaymentAttemptNew::default()
         };
-        state
-            .store
+        let store = state
+            .stores
+            .get(state.conf.multitenancy.get_tenant_names().first().unwrap())
```
[the remark this answers is not in the document]
Comment (jagan-jaya) [reply crates/router/src/types/storage/payment_attempt.rs:173]: used only in tests

File: crates/router/src/types/storage/payment_attempt.rs [hunk 8]
```diff
@@ -218,14 +223,16 @@ mod tests {
             attempt_id: uuid.clone(),
             ..PaymentAttemptNew::default()
         };
-        state
-            .store
+        let store = state
+            .stores
+            .get(state.conf.multitenancy.get_tenant_names().first().unwrap())
+            .unwrap();
```
[the remark this answers is not in the document]
Comment (jagan-jaya) [reply crates/router/src/types/storage/payment_attempt.rs:229]: used only in tests

File: crates/redis_interface/src/commands.rs [hunk 9]
```diff
@@ -520,22 +533,34 @@ impl super::RedisConnectionPool {
...
-    {
+    pub async fn stream_get_length(&self, stream: &str) -> CustomResult<usize, errors::RedisError> {
         self.pool
-            .xlen(stream)
+            .xlen(self.add_prefix(stream))
             .await
             .change_context(errors::RedisError::GetLengthFailed)
     }
 
+    pub fn get_tenant_keys<K>(&self, keys: K) -> MultipleKeys
```
[the remark this answers is not in the document]
Comment (jagan-jaya) [reply crates/redis_interface/src/commands.rs]: Renamed the method

**Review by jarnura — DISMISSED** [review]

**Review by NishantJoshi00 — APPROVED** [review]

**Review by jarnura — DISMISSED** [review]

**Review by lsampras — DISMISSED** [review]
analytics changes look good

**Review by jarnura — DISMISSED** [review]

**Review by Aprabhat19 — DISMISSED** [review]

**Review by lsampras — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Review by lsampras — APPROVED** [review]

**Review by Aprabhat19 — APPROVED** [review]

File: crates/router/src/services/api.rs [hunk 10]
```diff
@@ -915,15 +917,16 @@ pub enum AuthFlow {
 pub async fn server_wrap_util<'a, 'b, U, T, Q, F, Fut, E, OErr>(
     flow: &'a impl router_env::types::FlowMetric,
     state: web::Data<AppState>,
+    incoming_request_header: &HeaderMap,
```
Comment (lsampras) [inline crates/router/src/services/api.rs:920]: We should skip this else the request headers will be logged by tracing

**Pull request merged**